packages/react-components/src/gallery/dominantSpeaker.ts

```diff
@@ -0,0 +1,61 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+import { VideoGalleryRemoteParticipant } from '../types';
+
+/**
+ * Calculates the participants that should be renderd basis on the list of dominant speakers and currently renderd participants in a call.
+ * @param participants - Array containing all participants of a call. {@link @azure/communication-react#VideoGalleryRemoteParticipant}
+ * @param dominantSpeakers - An array containing the userId of dominant speakers in a call in the order of their dominance. 0th index is the most dominant, 1st is the second most etc
+ * @param visibleParticipants - Array containing currently rendered (visible) participants in the call. {@link @azure/communication-react#VideoGalleryRemoteParticipant}
+ * @param MAX_TILES - Maximum number of tiles to calculate.
+ * @returns VideoGalleryRemoteParticipant[] {@link @azure/communication-react#VideoGalleryRemoteParticipant}
+ */
+export const smartDominantSpeakerParticipants = (
+  participants: VideoGalleryRemoteParticipant[] = [],
+  dominantSpeakers: Array<string> = [],
+  visibleParticipants: VideoGalleryRemoteParticipant[] = [],
+  MAX_TILES = 4
+): VideoGalleryRemoteParticipant[] => {
+  if (!participants) return [];
+
+  // Only use the Max allowed dominant speakers.
+  const dominantSpeakerIds = dominantSpeakers.slice(0, MAX_TILES);
+
+  // Don't apply any logic if total number of video streams is less than Max video streams.
+  if (participants.length <= MAX_TILES) return participants;
+
+  // Initialize `visibleParticipants` if it is empty.
+  if (!visibleParticipants.length) {
+    visibleParticipants = participants.slice(0, MAX_TILES);
+  }
+
+  const lastVisibleSpeakerIds = visibleParticipants.map((speaker) => speaker.userId);
+  const newDominantSpeakerIds = dominantSpeakerIds.filter((id) => !lastVisibleSpeakerIds.includes(id));
+
+  // Remove participants that are no longer dominant and replace them with new dominant speakers.
+  lastVisibleSpeakerIds.forEach((id, idx) => {
```

Comment:
This updates `lastVisibleSpeakerIds`. This is a surprising side-effect of calling `smartDominantSpeakerParticipants`. I'd recommend treating input args as immutable. The caller of `smartDominantSpeakerParticipants` should explicitly assign the returned value to their stored `lastVisibleSpeakerIds` in order to be able to use it on next render.

Reply:
In order to avoid the caller having to generate the speaker IDs from the returned participant, this function could just accept the same type as the return: `previouslyVisibleParticipants: VideoGalleryRemoteParticipant[]`, and then get the IDs here as needed.

The caller is expected to do something like:

```
const previouslyVisibleParticipants = useRef([] as VideoGalleryRemoteParticipant[]);

...

const newVisibleParticipants = smartDominantSpeakerParticipants(participants, ..., previouslyVisibleParticipants.current);
previouslyVisibleParticipants.current = newVisibleParticipants;

// now render newVisibleParticipants

```

Reply:
- `lastVisibleSpeakerIds` is a local variable inside the function.  I think you are talking about `visibleParticipants`
- When this is implemented, video gallery will need to consume the participants returned by this function, so can't return just ID's as then video gallery will have to convert those IDs to participant again...

Reply:
Do read the code block again, I think you might have had a confusion.